client/src/quic_client.rs

```diff
@@ -0,0 +1,229 @@
+                client.runtime.block_on(send_buffer)?;
+                Ok(())
+            }
+        }
+    }
+
+    pub fn send_batch(&self, transactions: Vec<Transaction>) -> TransportResult<()> {
+        let buffers = transactions
+            .into_par_iter()
+            .map(|tx| bincode::serialize(&tx).expect("serialize Transaction in send_batch"))
+            .collect::<Vec<_>>();
+
+        match self {
+            Self::Udp(socket, addr) => {
+                buffers
+                    .into_iter()
+                    .try_for_each(|buff| -> TransportResult<()> {
+                        socket.send_to(&buff[..], addr)?;
+                        Ok(())
+                    })?;
+                Ok(())
+            }
+            Self::Quic(client) => {
+                let slices = buffers.par_iter().map(|buf| &buf[..]).collect::<Vec<_>>();
+                let _guard = client.runtime.enter();
+                let send_batch = client.send_batch(slices);
+                client.runtime.block_on(send_batch)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+impl QuicClient {
+    pub fn new(client_socket: UdpSocket, addr: SocketAddr) -> Self {
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let _guard = runtime.enter();
+
+        let crypto = rustls::ClientConfig::builder()
+            .with_safe_defaults()
+            .with_custom_certificate_verifier(SkipServerVerification::new())
+            .with_no_client_auth();
+
+        let create_endpoint = QuicClient::create_endpoint(EndpointConfig::default(), client_socket);
+
+        let mut endpoint = runtime.block_on(create_endpoint);
+
+        endpoint.set_default_client_config(ClientConfig::new(Arc::new(crypto)));
+
+        Self {
+            runtime,
+            endpoint,
+            connection: Arc::new(Mutex::new(None)),
+            addr,
+        }
+    }
+
+    async fn create_endpoint(config: EndpointConfig, client_socket: UdpSocket) -> Endpoint {
+        quinn::Endpoint::new(config, None, client_socket).unwrap().0
+    }
+
+    async fn _send_buffer_using_conn(
+        data: &[u8],
+        connection: &Arc<NewConnection>,
+    ) -> Result<(), WriteError> {
+        let mut send_stream = connection.connection.open_uni().await?;
+        send_stream.write_all(data).await?;
+        send_stream.finish().await?;
+        Ok(())
+    }
+
+    // Attempts to send data, connecting/reconnecting as necessary
+    // On success, returns the connection used to successfully send the data
+    pub async fn send_buffer(&self, data: &[u8]) -> Result<Arc<NewConnection>, WriteError> {
+        let connection = {
+            let mut conn_guard = self.connection.lock().await;
+
+            let maybe_conn = (*conn_guard).clone();
+            match maybe_conn {
+                Some(conn) => conn.clone(),
+                None => {
```

Comment:
Am I right in the assumption that there is no contention for the mutex in this case?

Reply:
Yeah, the idea is that assuming we're only getting one or a few concurrent calls into batch_send at a time, there shouldn't be much lock contention since each batch send only touches the lock once or twice.